web/packages/teleport/src/AppLauncher/AppLauncher.tsx

```diff
@@ -41,6 +41,21 @@ export function AppLauncher() {
     const port = location.port ? `:${location.port}` : '';
 
     try {
+      // Attempt to resolve the fqdn of the app, if we can't then an error
+      // will be returned preventing a redirect to a potentially arbitrary
+      // address. Compare the resolved fqdn with the one that was passed,
+      // if they don't match then the public address was used to find the
+      // resolved fqdn, and the passed fdqn isn't valid.
+      const resolvedFqdn = await service.getAppFqdn({
+        fqdn: params.fqdn,
+        clusterId: params.clusterId,
+        publicAddr: params.publicAddr,
+        arn: params.arn,
+      });
+      if (resolvedFqdn.fqdn != params.fqdn) {
+        throw Error(`Failed to match applications with FQDN ${params.fqdn}`);
+      }
```

Comment:
hrm, i feel like this check should be done at the backend because the backend states:

```
// getAppFQDN resolves the input params to a known application and returns
// its valid FQDN.
```

otherwise it makes that ^ statement inaccurate, we can resolve the app, but the fqdn might be invalid